### PR TITLE
VB-3443 Add automation tests for do not change

### PR DESCRIPTION
**Pull request by AledWEvans**
Guys a lot of changes here,.. but have the do not change tickets running, please have a look.
Need to rase more tickets next week to fix other issues.

### CODE REVIEW

File: src/test/java/uk/gov/justice/digital/hmpps/vsip/services/PrisonVisitsTestingHelperService.java [hunk 1]
```diff
@@ -76,14 +78,35 @@ public void startPrisonerRestriction(String prisonerCode) {
...
-            // We use changing here because it is not counted as part of booking slot counts and is also removed after configured time.
-            bookings.forEach(bookingReference -> client.changeStatus(bookingReference, VisitStatus.CHANGING));
+    public void cleanUpBookingsAndApplications() {
+        LOG.debug("Entered cleanUpLastBooking"
+        );
+        if (context.getBookingReferences()!=null) {
+            var bookings = new ArrayList<>(context.getBookingReferences());
+            if (bookings != null) {
+                LOG.debug(bookings.size() + " Bookings found, must tidy up!");
+                // We use changing here because it is not counted as part of booking slot counts and is also removed after configured time.
```
Comment (tpmcgowan) [inline src/test/java/uk/gov/justice/digital/hmpps/vsip/services/PrisonVisitsTestingHelperService.java:88]: Don't think this comment is still accurate as `CHANGING` status no longer relevant?

File: src/test/java/uk/gov/justice/digital/hmpps/vsip/services/PrisonVisitsTestingHelperService.java [hunk 2]
```diff
@@ -76,14 +78,35 @@ public void startPrisonerRestriction(String prisonerCode) {
...
+                    context.getBookingReferences().remove(bookingReference);
+                });
+            }
         }
+
+        if (context.getApplicationReferences()!=null) {
+            var applicationReferences = new ArrayList<>(context.getApplicationReferences());
+            if (applicationReferences != null) {
+                LOG.debug(applicationReferences.size() + " Application found, must tidy up!");
+                // We use changing here because it is not counted as part of booking slot counts and is also removed after configured time.
```
Comment (tpmcgowan) [inline src/test/java/uk/gov/justice/digital/hmpps/vsip/services/PrisonVisitsTestingHelperService.java:101]: Same as above - `CHANGING`

File: src/test/java/uk/gov/justice/digital/hmpps/vsip/cucumber/steps/ViewVisitsByDateStep.java [hunk 3]
```diff
@@ -47,4 +58,29 @@ public void iSeeOnViewVisitsByDatePage(String errMsg) {
         viewVisitsByDatePage.checkErrorMessage();
     }
 
+    @And("I capture date the Booking is booked")
+    public void iCaptureDateTheBookingIsBooked() {
+        //capturing date from booking confirmation page
+        String originalDateValue = confirmationPage.getDateBooked();
+        System.out.println("Original Date value from confirmation page: " + originalDateValue);
```
Comment (tpmcgowan) [inline src/test/java/uk/gov/justice/digital/hmpps/vsip/cucumber/steps/ViewVisitsByDateStep.java:65]: Should these two `System.out`s be `LOG`?

**Review by tpmcgowan — APPROVED** [review]
LGTM - just a few very minor comments

File: src/test/java/uk/gov/justice/digital/hmpps/vsip/pages/HomePage.java [hunk 4]
```diff
@@ -44,8 +50,25 @@ public void gotoHomePage() {
...
 
-    public void hashNeedReviewCount() {
+    public void hasNeedReviewCount() {
+        findNeedReviewCount.isDisplayed();
         String count = findNeedReviewCount.getText();
-        Assert.assertTrue("Count should be greater than 0 value (" + count + ")", Integer.parseInt(count.trim()) > 0);
+        Assert.assertTrue("Count should be greater than 0 value (" + count + ")", isPositiveNumber(count));
+    }
+
+    public static boolean isPositiveNumber(String strNum) {
```
Comment (dhirajshettymoj) [inline src/test/java/uk/gov/justice/digital/hmpps/vsip/pages/HomePage.java:59]: isANumber?

File: src/test/java/uk/gov/justice/digital/hmpps/vsip/cucumber/steps/ViewVisitsByDateStep.java [hunk 5]
```diff
@@ -47,4 +58,29 @@ public void iSeeOnViewVisitsByDatePage(String errMsg) {
...
+        System.out.println("Original Date Booking Reference: " + testContextService.getOriginalDateBookingReference());
+
+        //formatting the date to pass to API
+        DateTimeFormatter originalFormatter = DateTimeFormatter.ofPattern("EEEE d MMMM yyyy");
+        // Parse the original date string
+        LocalDate originalDate = LocalDate.parse(originalDateValue, originalFormatter);
+        // Define the formatter for the desired output format
+        DateTimeFormatter outputFormatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");
+        // Format the date to the desired output format
+        String formattedDate = originalDate.format(outputFormatter);
```
Comment (dhirajshettymoj) [inline src/test/java/uk/gov/justice/digital/hmpps/vsip/cucumber/steps/ViewVisitsByDateStep.java:77]: should we move these to a getFormattedDate method?

File: src/test/java/uk/gov/justice/digital/hmpps/vsip/cucumber/steps/CancelAVisitSteps.java [hunk 6]
```diff
@@ -25,6 +31,36 @@ public void clickOnCancelButton() {
...
+    public void selectYesIAmSureTheVisitDoseNotNeedToBeUpdatedOrCancelled() {
+        cancelAVisitPage.clickSelectYesIAmSureTheVisitDoseNotNeed();
+    }
+    @And("I select no i am sure the visit dose not need to be updated or cancelled")
+    public void iSelectNoIAmSureTheVisitDoseNotNeedToBeUpdatedOrCancelled() {
+        cancelAVisitPage.clickSelectNoIAmSureTheVisitDoseNotNeed();
+    }
+
+    @And("I enter a {string} as a reason why the visit dose not need to be updated or cancelled")
+    public void iEnterAReasonWhyTheVisitDoseNotNeedToBeUpdatedOrCancelled(String reason) {
```
Comment (dhirajshettymoj) [inline src/test/java/uk/gov/justice/digital/hmpps/vsip/cucumber/steps/CancelAVisitSteps.java:60]: minor typo - does

**Review by dhirajshettymoj — COMMENTED** [review]
some minor comments added. looks good overall!

File: src/test/java/uk/gov/justice/digital/hmpps/vsip/cucumber/steps/ViewVisitsByDateStep.java (hunk 5 above)
Comment (AledWEvans) [reply src/test/java/uk/gov/justice/digital/hmpps/vsip/cucumber/steps/ViewVisitsByDateStep.java:81]: Not my code will need to be fixed up later, lots more work to be done

**Review by dhirajshettymoj — APPROVED** [review]
LGTM

**Review by tpmcgowan — APPROVED** [review]
LGTM

**Pull request merged**